### PR TITLE
Some improvements to validation webhook

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[x] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pkg/webhooks/validation/controller/controller.go [hunk 1]
```diff
@@ -225,11 +225,11 @@ func newController(
 	webhookInformer := cache.NewSharedIndexInformer(
 		&cache.ListWatch{
 			ListFunc: func(opts metav1.ListOptions) (runtime.Object, error) {
-				opts.LabelSelector = fmt.Sprintf("%s=%s", label.IoIstioRev.Name, o.Revision)
+				opts.FieldSelector = fields.OneTermEqualSelector("metadata.name", c.webhookName).String()
```
Comment (howardjohn) [inline pkg/webhooks/validation/controller/controller.go:228]: I think the old code is intentionally to handle multiple webhooks? @Monkeyanator

**Review by howardjohn — COMMENTED** [review]
Can you explain what is being improved? Since there are no tests changes its hard to tell if its refactoring, bug fix, etc

File: pkg/webhooks/validation/controller/controller.go (hunk 1 above)
Comment (Monkeyanator) [reply pkg/webhooks/validation/controller/controller.go:228]: Should be fine just targeting the single webhook, nothing relies on being able to handle multiple at the moment.

File: pkg/webhooks/validation/controller/controller.go [hunk 2]
```diff
@@ -243,19 +243,13 @@ func newController(
...
 	go c.webhookInformer.Run(stop)
 	if !cache.WaitForCacheSync(stop, c.webhookInformer.HasSynced) {
 		return
 	}
 	go c.startCaBundleWatcher(stop)
 	go c.runWorker()
-
 	<-stop
-	c.queue.Add(&reconcileRequest{
-		event:       quitEvent,
```
Comment (hzxuzhonghu) [inline pkg/webhooks/validation/controller/controller.go:256]: cleanups

File: pkg/webhooks/validation/controller/controller.go [hunk 3]
```diff
@@ -319,31 +303,34 @@ func (c *Controller) processNextWorkItem() (cont bool) {
 // reconcile the desired state with the kube-apiserver.
 // the returned results indicate if the reconciliation should be retried and/or
 // if there was an error.
-func (c *Controller) reconcileRequest(req *reconcileRequest) (bool, error) {
+func (c *Controller) reconcileRequest(req *reconcileRequest) error {
```
Comment (hzxuzhonghu) [inline pkg/webhooks/validation/controller/controller.go:306]: simplify the returns value processing, previously it is too complex

File: pkg/webhooks/validation/controller/controller.go [hunk 4]
```diff
@@ -243,19 +243,13 @@ func newController(
 
 func (c *Controller) Run(stop <-chan struct{}) {
 	defer c.queue.ShutDown()
-	c.client.RunAndWait(stop)
```
Comment (hzxuzhonghu) [inline pkg/webhooks/validation/controller/controller.go:246]: unnecessary calling

File: pkg/webhooks/validation/server/server.go [hunk 5]
```diff
@@ -120,23 +120,6 @@ func New(o Options) (*Webhook, error) {
...
 }
 
-// Stop the server
-func (wh *Webhook) Stop() {
-}
-
-var readyHook = func() {}
-
-// Run implements the webhook server
-func (wh *Webhook) Run(stopCh <-chan struct{}) {
```
Comment (hzxuzhonghu) [inline pkg/webhooks/validation/server/server.go:130]: cleanup

**Comment by hzxuzhonghu** [comment]
This is more of a cleanup and also an optimization

File: pkg/webhooks/validation/controller/controller.go (hunk 1 above)
Comment (howardjohn) [reply pkg/webhooks/validation/controller/controller.go:228]: I am not sure I understand, we install two now:
```
$ istioctl manifest generate | k grep 'Validat*/' -s
ValidatingWebhookConfiguration/istiod.
ValidatingWebhookConfiguration/istiod-istio-system.
```

So why do we only need to watch one?

File: pilot/pkg/bootstrap/validation.go [hunk 6]
```diff
@@ -33,16 +33,11 @@ func (s *Server) initConfigValidation(args *PilotArgs) error {
...
 		Mux:          s.httpsMux,
 	}
-	whServer, err := server.New(params)
+	_, err := server.New(params)
 	if err != nil {
 		return err
 	}
 
-	s.addStartFunc(func(stop <-chan struct{}) error {
-		whServer.Run(stop)
```
Comment (howardjohn) [inline pilot/pkg/bootstrap/validation.go:42]: where is this called now?

Comment (howardjohn) [reply pilot/pkg/bootstrap/validation.go:42]: Oh you deleted it. I got confused with NewValidatingWebhookController.Run()

File: pkg/webhooks/validation/controller/controller.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pkg/webhooks/validation/controller/controller.go:228]: Here  in the controller, it only reconciles one validatingwebhookconfiguration.

Comment (hzxuzhonghu) [reply pkg/webhooks/validation/controller/controller.go:228]: The name is fixed

Comment (howardjohn) [reply pkg/webhooks/validation/controller/controller.go:228]: I understand it is today. But I don't think it should be. If we install 2, we need to reconcile 2.

So we need to either install one, or reconcile both?

Comment (hzxuzhonghu) [reply pkg/webhooks/validation/controller/controller.go:228]: I would argue each istiod should take charge of reconciling its own webhookconfiguration. Why does it need to reconcile another revision?

Comment (hzxuzhonghu) [reply pkg/webhooks/validation/controller/controller.go:228]: OK, tried. I will revert this change. And I am sure we want to clean up in the future.

**Review by howardjohn — CHANGES_REQUESTED** [review]
When we fail to update the webhook this never retries anymore:
```
2021-06-17T15:20:44.845092Z     info    validationController    Reconcile(enter): [description] CA bundle update, [eventType] updateEvent
2021-06-17T15:20:44.848889Z     info    validationController    Not ready to switch validation to fail-closed: dummy invalid config not rejected
2021-06-17T15:20:44.849095Z     info    validationController    validatingwebhookconfiguration istiod (failurePolicy=Ignore, resourceVersion=155087) is up-to-date. No change required.
... no logs...
```

I think we need some more tests here

**Comment by hzxuzhonghu** [comment]
ping @howardjohn left only some small changes

**Review by howardjohn — CHANGES_REQUESTED** [review]
The retries are still not right. They used to have exponential backoff which has been removed. This is important since before we saw folks with $10k logging bills from all the attempts...

Since its been broken 2-3 times now, I don't feel comfortable approving any changes that don't extend the test coverage of this..

**Comment by hzxuzhonghu** [comment]
Rate limited queue has a backoff retry natively, i am not sure what you mean.

**Comment by howardjohn** [comment]
if you go try it you will see it runs every second

**Comment by hzxuzhonghu** [comment]
can you point where do I modify the queue? BTW, i will have a test

**Comment by hzxuzhonghu** [comment]
```
root@opensrc:/mnt/go/src/istio.io/istio# grep validationController log
2021-06-23T02:44:57.419256Z	info	validationController	Reconcile(enter): [description] add event (admissionregistration.k8s.io/v1, Kind=ValidatingWebhookConfiguration) istio-validator-istio-system, [eventType] updateEvent
2021-06-23T02:44:58.423724Z	info	validationController	Not ready to switch validation to fail-closed: dummy invalid rejected for the wrong reason: Internal error occurred: failed calling webhook "validation.istio.io": Post "https://istiod.istio-system.svc:443/validate?timeout=10s": dial tcp 10.96.158.254:443: connect: connection refused
2021-06-23T02:44:58.428461Z	info	validationController	Successfully updated validatingwebhookconfiguration istio-validator-istio-system (failurePolicy=Ignore,resourceVersion=14758318)
2021-06-23T02:44:58.431085Z	info	validationController	Reconcile(enter): [description] CA bundle update, [eventType] updateEvent
2021-06-23T02:44:59.435765Z	info	validationController	Not ready to switch validation to fail-closed: dummy invalid config not rejected
2021-06-23T02:44:59.436109Z	info	validationController	validatingwebhookconfiguration istio-validator-istio-system (failurePolicy=Ignore, resourceVersion=14758318) is up-to-date. No change required.
2021-06-23T02:44:59.438131Z	info	validationController	Reconcile(enter): [description] update event (admissionregistration.k8s.io/v1, Kind=ValidatingWebhookConfiguration) istio-validator-istio-system, [eventType] updateEvent
2021-06-23T02:45:00.449507Z	info	validationController	Endpoint successfully rejected invalid config. Switching to fail-close.
2021-06-23T02:45:00.454543Z	info	validationController	Successfully updated validatingwebhookconfiguration istio-validator-istio-system (failurePolicy=Fail,resourceVersion=14758333)
2021-06-23T02:45:00.457221Z	info	validationController	Reconcile(enter): [description] retry reconcile request, [eventType] retryEvent
2021-06-23T02:45:00.457994Z	info	validationController	validatingwebhookconfiguration istio-validator-istio-system (failurePolicy=Fail, resourceVersion=14758333) is up-to-date. No change required.
2021-06-23T02:45:00.460010Z	info	validationController	Reconcile(enter): [description] retry reconcile request, [eventType] retryEvent
2021-06-23T02:45:00.460351Z	info	validationController	validatingwebhookconfiguration istio-validator-istio-system (failurePolicy=Fail, resourceVersion=14758333) is up-to-date. No change required.
2021-06-23T02:45:00.462369Z	info	validationController	Reconcile(enter): [description] update event (admissionregistration.k8s.io/v1, Kind=ValidatingWebhookConfiguration) istio-validator-istio-system, [eventType] updateEvent
2021-06-23T02:45:00.462688Z	info	validationController	validatingwebhookconfiguration istio-validator-istio-system (failurePolicy=Fail, resourceVersion=14758333) is up-to-date. No change required.

```

This is the log, not sure how you tested. 

Since i did not change the queue related logic, i am not sure what test coverage you want

**Comment by howardjohn** [comment]
Yeah that shows the problem - retest every 1s. It's supposed to be an exponential backoff on the retry

**Comment by hzxuzhonghu** [comment]
hmm, they are triggered by different events, so the items enqueue is not same, so the rate limit queue does not limit. 
The rate limit applies to a signle key

**Comment by hzxuzhonghu** [comment]
@howardjohn If no other comments, can you approve

**Comment by howardjohn** [comment]
It doesn't look like the retry backoff is fixed?

**Comment by hzxuzhonghu** [comment]
@howardjohn Can you point to where  the problem is,  i think I explained that above https://github.com/istio/istio/pull/33315#issuecomment-866488750/

**Comment by howardjohn** [comment]
I am fairly sure before this PR it had backoff. Doug added it a few weeks
back. Unless it broke sometime between a few weeks ago and now, I didn't
test on master I just tried this PR and saw it's not backing off anymore.

**Comment by hzxuzhonghu** [comment]
So you donot have evidence where it is wrong, just a guess. I feel really uncomfortable continuing with this, anyone can take it over.

**Comment by hzxuzhonghu** [comment]
Isn't this real hostile as @Stono said?

**Comment by howardjohn** [comment]
@hzxuzhonghu I mean nothing hostile whatsoever. To the best if my knowledge this is reverting the changes in https://github.com/istio/istio/pull/32210 which I feel are an important bug fix. It is unfortunate that the original PR didn't have any tests to catch the regression.

Sorry you don't feel comfortable continuing with this, I think it was a good improvement aside from the regression on the retry backoff

**Comment by howardjohn** [comment]
@hzxuzhonghu my apologies for wasting your time here. I had thought this was breaking the recent bug fix mentioned in https://github.com/istio/istio/pull/33315#issuecomment-868927604, but I only tested your PR. It turns out its broken in `master` as well, and your PR didn't make it any worse. Sorry!!

**Review by howardjohn — APPROVED** [review]

**Comment by howardjohn** [comment]
I opened https://github.com/istio/istio/pull/33676 as a followup

**Comment by Monkeyanator** [comment]
@hzxuzhonghu just realized from these comments and the followup that my PR #33044 broke the retries. Sorry I lost track of the discussion here and caused you unnecessary trouble.

**Comment by hzxuzhonghu** [comment]
All right, we are all constructing a better project and a better community.

**Pull request merged**